### PR TITLE
Shopping cart

### PR DESCRIPTION
**Pull request by jklemen**

### CODE REVIEW

File: shopping_cart/features/account_login.feature [hunk 1]
```diff
@@ -0,0 +1,49 @@
+#encoding: utf-8
```
Comment (jmmastey) [inline shopping_cart/features/account_login.feature]: ??

File: shopping_cart/features/account_login.feature [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  As an online shopper
+  I should enter valid credentials
+
+  @happy
+  Scenario: I log in while in a guest session
+    Given I have an open session with items in my cart
+    And I am not logged in
+    And I have an account on the shopping site
+    When I log in
+    Then my items will still be in my shopping cart
```
Comment (jmmastey) [inline shopping_cart/features/account_login.feature]: This is somewhat vague. Might be better as:

```
Given I am a guest with 2 items in my cart
When I log in as a registered user
Then I have 2 items in my cart
```

File: shopping_cart/features/account_login.feature [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+  @happy
+  Scenario: I log out while in a registered user session
+    Given I have an open session with items in my cart
+    And I am logged in
+    When I log out
+    Then my my cart should be empty
```
Comment (jmmastey) [inline shopping_cart/features/account_login.feature]: Similar:

```
Given I am a registered user with 2 items in my cart
When I log out
Then my cart is empty
```

File: shopping_cart/features/account_login.feature [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+  @happy
+  Scenario: Merging carts
+    Given I have two open sessions with different items in each cart
+    And I am not logged in on either session
+    And I have an account on the shopping site
```
Comment (jmmastey) [inline shopping_cart/features/account_login.feature]: This step shows up consistently, so it belongs in `Background`, not each test.

File: shopping_cart/features/account_login.feature [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    When I log into both sessions
+    Then my items from both sessions will merge into my login cart
+
+  @happy
+  Scenario: I enter valid credentials
+    Given I have an account on the shopping site
+    When I correctly enter my username and password
+    Then I will be taken to the home page
```
Comment (jmmastey) [inline shopping_cart/features/account_login.feature]: `will be` is odd phrasing.

```
When I enter my correct username and password
Then I am on the homepage
And I see a message that I have signed in successfully.
```

File: shopping_cart/features/address_entry.feature [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    And I'm trying to get a shipping estimate
+    And I have entered an invalid street, city, state, country or zip
+    Then an invalid address component error should appear
+    And tell me which field(s) that i have entered are invalid
+
+  @bad
+  Scenario: Address field(s) are blank
+    Given I'm at the online store
+    And I'm trying to get a shipping estimate
+    And I have entered nothing in 1 or more fields
```
Comment (jmmastey) [inline shopping_cart/features/address_entry.feature]: This isn't something you'd probably write a test against. More likely you'd have a specific `And I leave my last name blank` or similar

File: shopping_cart/features/coupon_add.feature [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    And I click checkout
+    Then the coupon should be un-applied
+
+  @happy
+  Scenario: Coupon is valid for one item
+    Given I'm at the online store
+    When I enter a coupon code valid for a single item in my cart
+    Then the price of that item should change
+    And the price of no other item of its type should change
+    And the total of my shopping cart should update
```
Comment (jmmastey) [inline shopping_cart/features/coupon_add.feature]: These benefit from literal tests:

```
Given I have two fire hoses in my cart for $20
When I enter a coupon for $5 off
Then I should see that one fire hose is discounted
And my cart subtotal is $15
```

File: shopping_cart/features/modify_cart.feature [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    And my total price should subtract the item's price
+
+  @happy
+  Scenario: Removing an item from my shopping cart
+    Given I'm at the online store
+    And I have an item in my shopping cart
+    When I press delete on the item
+    Then the item should be removed from my shopping cart
+    And my total price should subtract the item's prices
+
```
Comment (jmmastey) [inline shopping_cart/features/modify_cart.feature]: What if I add more than are in stock?

File: shopping_cart/features/shipping_estimates.feature [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    And I have at least 1 item in my shopping cart
+    When I try to get a shipping estimate
+    And my address is a <address_type>
+    Then the estimate should be <cost>
+
+  Examples:
+    | address_type  | cost |
+    | house         | 50   |
+    | apartment     | 100  |
+    | business      | 200  |
```
Comment (jmmastey) [inline shopping_cart/features/shipping_estimates.feature]: Nice use of example groups

**Comment by jmmastey** [comment]
Looks good. Take a look at the feedback, most of it is about phrasing of your test cases. Remember that reading the feature files is the most common way people will onboard into your project, so their clarity is super important.

I found one missing case, but that's not bad at all :d 

Move along!

File: shopping_cart/features/account_login.feature (hunk 1 above)
Comment (jklemen) [inline shopping_cart/features/account_login.feature]: I thought this line was necessary from a tutorial I did >_< sorry

File: shopping_cart/features/account_login.feature (hunk 4 above)
Comment (jklemen) [inline shopping_cart/features/account_login.feature]: Just learned about Background today! Will change.

File: shopping_cart/features/modify_cart.feature (hunk 8 above)
Comment (jklemen) [inline shopping_cart/features/modify_cart.feature]: Then my poor online store would have no choice but to attempt to process that. Good catch

File: shopping_cart/features/coupon_add.feature (hunk 7 above)
Comment (jklemen) [inline shopping_cart/features/coupon_add.feature]: Good to know!

**Pull request closed without merging**